mono/metadata/object.c

```diff
@@ -5305,6 +5305,58 @@ mono_runtime_try_invoke_array (MonoMethod *method, void *obj, MonoArray *params,
 	}
 }
 
+/**
+ * mono_object_new_common_tail:
+ *
+ * This function centralizes post-processing of objects upon creation.
+ * i.e. calling mono_object_register_finalizer and mono_gc_register_obj_with_weak_fields,
+ * and setting error.
+ */
+static MonoObject *
+mono_object_new_common_tail (MonoObject* o, MonoClass *klass, MonoError* error)
+{
+	error_init (error);
+
+	if (G_UNLIKELY (!o)) {
+		mono_error_set_out_of_memory (error, "Could not allocate %i bytes", klass->instance_size);
+		return o;
+	}
+
+	if (klass->has_finalize)
+		mono_object_register_finalizer (o);
+
+	if (klass->has_weak_fields)
+		mono_gc_register_obj_with_weak_fields (o);
+
+	return o;
+}
+
+#if 0 // FIXMEcoop awaiting https://github.com/mono/mono/pull/6876
+/**
+ * mono_object_new_handle_tail:
+ *
+ * This function centralizes post-processing of objects upon creation.
+ * i.e. calling mono_object_register_finalizer and mono_gc_register_obj_with_weak_fields.
+ */
+static MonoObjectHandle
+mono_object_new_handle_common_tail (MonoObjectHandle o, MonoClass *klass, MonoError* error)
```

Comment:
Handle'izing the guts of object allocation might not be very productive.

Reply:
I think, if I understand, correctly, Ludovic wants like this:

 https://github.com/mono/mono/pull/6876

and then start converting to it -- bottom-up coop/handle conversion.

He asked for the lowest levels of that, which I think implies the next level, which implies the unfinished part of this PR. I'm open to a few things or more -- this PR can ignore handles, or it can be incomplete as shown, and/or we we can discuss the other one more. I understand you can layer new_handle over existing raw creators but I think Ludovic implied a desire to push handle_new down. Please advise.

Eventually maybe hypothetically you change existing raw-creating-names to be handle-creating, thereby achieving less churn. There is some ability maybe to do that with C++, except you can't overload on return type, so maybe not. Actually, I have a theory that we should be returning errors-as-ints everywhere (or MonoError by rvalue), and then 'actual' return values are 'out parameters', and then overloading works, voila..but I not sure how soon we will have C++, so overloading cannot apply.) This notion of "losing the return value to errors and using out parameters extensively" is pervasive in some code bases. It seems unfortunate, but is the price for reliable error checking in exception-free code.

So, absent all that hypothetical and longer term stuff, we provide handle-based overloads with new names and work them through the entire system.